### ISSUE TITLE
Adding a InventoryChange event to improve performance

### ISSUE THREAD

**Issue by WindBlades** [MinecraftForge/MinecraftForge#6255]
A lot of mods need to know when an item is added in the player inventory or removed (from any source like pickup or remove/add from a contained), most of them just check every ticks and it's pretty heavy in performance wise.

So why not just add an event that add a hook like minecraft do for their achievement with InventoryChangeTrigger.
What are your thoughts on that ?

**Comment by diesieben07** [comment]
> it's pretty heavy in performance wise.

proof? Just waving your hands and saying "performance!" does not mean there is a performance problem.

**Comment by WindBlades** [comment]
> proof? Just waving your hands and saying "performance!" does not mean there is a performance problem.

Well, i'm sure you can see why having a multiple mods checking 20 time by second if each slot of an inventory have changed and that for each items they want to check is way heavier compared to a hook that will be trigerred only one time when a slot change.
Minecraft have choose the hook for the achievement instead of the tick one.

**Comment by pau101** [comment]
There is inherently a performance concern with mechanics of the game that are written which reply on tick-based conditional logic for detecting and acting upon state changes. Issues of this kind are exacerbated by any number of mods which add content of this sort. In order to remedy this concern, designs which introduce pub-sub and observer based patterns reduce the tick overhead by removal of mass code. The implementation of a potential solution is where performance metrics guide finding the most practical yet optimized design.

**Comment by williewillus** [comment]
yes, but forge isn't and can't be in the business of making large modifications of the game's core design aspects like that.

**Comment by pau101** [comment]
I am not suggesting making any modifications to the game's code more than the patches required to add a typical hook. One event which already exists which is similar to this proposed idea is `LivingEquipmentChangeEvent`.

**Comment by raoulvdberge** [comment]
The same argument can be applied to any inventory, not only player inventories.

**Comment by Unknown303YT** [comment]
This issue is still relevant, as I need to detect when the player's inventory is changed.

**Comment by Jonathing** [comment]
Ok. We could make a `PlayerContainerEvent.Modified` and feed the same context that is given to the `CriteriaTriggers.INVENTORY_CHANGED` criteria (see `ServerPlayer.containerListener`). This would be the player, the abstract container menu, the slot number, and the relevant item stack. Is there anything else you'd need or would that context be enough?

**Comment by Jonathing** [comment]
After giving it some more thought, I think an additional event is unnecessary. By listening to `PlayerContainerEvent.Open`, you can add your own container listener to the event. This saves Forge the trouble of firing an event every time a slot in the container is changed. Here's an example of how you can do this:

```java
@SubscribeEvent
public static void onPlayerContainerOpen(PlayerContainerEvent.Open event) {
    event.getContainer().addSlotListener(new ContainerListener() {
        @Override
        public void slotChanged(AbstractContainerMenu menu, int slotNumber, ItemStack stack) {
            Slot slot = menu.getSlot(slotNumber);

            // if this is a result slot, we probably don't care about the change
            if (slot instanceof ResultSlot) return;

            if (stack.isEmpty()) {
                LOGGER.info("Slot {} is now empty in {}", slotNumber, menu);
            }
        }

        @Override
        public void dataChanged(AbstractContainerMenu menu, int slotNumber, int data) { }
    });
}
```

So for 1.21.4, this is not an issue. For now I will keep this issue open as a reminder to myself to check older versions. If container listeners don't exist in a useful way in older versions, then `PlayerContainerEvent.Update` might be in order.

**Comment by Jonathing** [comment]
Closing this once again. I've double-checked all the way back to 1.18.2, and container listeners have existed since then. So, use the example I showed earlier to track when containers are updated.
